src/main/java/in/projecteka/consentmanager/user/SessionService.java

```diff
@@ -1,36 +1,72 @@
+import in.projecteka.consentmanager.user.model.OtpVerificationRequest;
+import in.projecteka.consentmanager.user.model.OtpVerificationResponse;
 import in.projecteka.consentmanager.user.model.SessionRequest;
 import lombok.AllArgsConstructor;
+import lombok.RequiredArgsConstructor;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.util.StringUtils;
 import reactor.core.publisher.Mono;
 
+import java.util.UUID;
+
 import static in.projecteka.consentmanager.common.Constants.BLACKLIST;
 import static in.projecteka.consentmanager.common.Constants.BLACKLIST_FORMAT;
 
+@RequiredArgsConstructor
 @AllArgsConstructor
 public class SessionService {
 
     private final TokenService tokenService;
     private final CacheAdapter<String, String> blacklistedTokens;
+    private CacheAdapter<String, String> unverifiedSessions;
     private final Logger logger = LoggerFactory.getLogger(SessionService.class);
+    private final UserRepository userRepository;
+    private final OtpServiceClient otpServiceClient;
+    private final OtpServiceProperties otpServiceProperties;
 
     public Mono<Session> forNew(SessionRequest request) {
         if (StringUtils.isEmpty(request.getUsername()) || StringUtils.isEmpty(request.getPassword()))
-            return Mono.error(ClientError.unAuthorizedRequest());
+            return Mono.error(ClientError.unAuthorizedRequest("Username or password is incorrect"));
         return tokenService.tokenForUser(request.getUsername(), request.getPassword())
                 .doOnError(error -> logger.error(error.getMessage(), error))
-                .onErrorResume(error -> Mono.error(ClientError.unAuthorizedRequest()));
+                .onErrorResume(error -> Mono.error(ClientError.unAuthorizedRequest("Username or password is incorrect")));
     }
 
     public Mono<Void> logout(String accessToken, LogoutRequest logoutRequest) {
         return blacklistedTokens.put(String.format(BLACKLIST_FORMAT, BLACKLIST, accessToken), "")
                 .then(tokenService.revoke(logoutRequest.getRefreshToken()));
     }
+
+    public Mono<OtpVerificationResponse> sendOtp(OtpVerificationRequest otpVerificationRequest) {
+        String sessionId = UUID.randomUUID().toString();
+        return userRepository.userWith(otpVerificationRequest.getUsername())
+                .switchIfEmpty(Mono.error(ClientError.userNotFound()))
+                .map(user -> new OtpCommunicationData("mobile",user.getPhone()))
+                .map(otpCommunicationData -> new OtpRequest(sessionId,otpCommunicationData))
+                .flatMap(requestBody -> otpServiceClient.send(requestBody)
+                        .then(Mono.defer(() -> unverifiedSessions.put(sessionId, otpVerificationRequest.getUsername())))
+                        .thenReturn(requestBody.getCommunication().getValue()))
+                .map(mobileNumber -> new OtpVerificationResponse(sessionId,mobileNumber,otpServiceProperties.getExpirationTime()));
```

Comment:
Shall we move the mobile number and expiry inside meta, contract also needs to be changed, sorry I did not notice it properly.